### ISSUE TITLE
How to get negative chips?

### ISSUE THREAD

**Issue by xiaoyongzhu** [mahyarnajibi/SNIPER#51]
Hi,

I am currently trying to improve my performance for a certain dataset, and I am not using negative chips to train my algorithms. From the visualization it looks fine, however there are some false positives that I do want to avoid. I am pretty sure that is because of the fact that I don't do negative chip training.

Can you specify more on how to get the negative chips? From the paper it says:
```
To this end, for negative chip mining, we first train RPN for a couple of epochs. No negative chips are used for training this network. The task of this network is to roughly guide us in selecting regions which are likely to contain false positives, so it is not necessary for it to be very accurate. This RPN is used to generate proposals over the entire training set. We assume that if no proposals are generated in a major portion of the image by RPN, then it is unlikely to contain an object instance.
```

And in the code, I found that there's a setting called `TEST.EXTRACT_PROPOSALS` - but reading thru the code, I am not sure if it is doing what's described in the paper. 

Any pointers will be appreciated! Thanks!

**Comment by bharatsingh430** [comment]
We will fix the repo by mid August. Last 2-3 weeks were a bit busy, so we didn’t update the repo, but the negative chip code and changing to resolution from scale so that it’s easy to train under new settings would be added. 

You can first train rpn just with positive samples and then extract proposals. Once that is done, try 2 or 3 negative chips per image (it’s a hyper parameter, so depends on the dataset) for negative mining.

**Comment by xiaoyongzhu** [comment]
Thanks for the response! I actually have another idea, since in the code there's a function called `extract_proposals`, maybe I can just reuse that for now. I don't think it's doing negative proposal mining specifically, however it will give a lot of proposals including both negative and positive ones, which might still be useful for the training.

I tried the above approach and didn't understand one thing - the `extract_proposals` function actually outputs the proposals for all the scales (in the COCO case, it's 3 scales). But I guess for the negative chips mining's purpose, should I just use a single scale (the base scale)?

Thanks!

**Comment by mahyarnajibi** [comment]
We used proposals extracted from all three scales for negative chip mining. Also, the code itself takes care of removing the already covered positive proposals and keeping only the remaining negative proposals by default. So once you extracted the proposals for your training set by setting the ```TRAIN.EXTRACT_PROPOSALS```, you can use them instead of the provided proposals.

**Comment by mySweetie** [comment]
have the negative chip mining code been released?thanks!

**Comment by xiaoyongzhu** [comment]
@mySweetie see above for my comments. I think you should set `TEST.EXTRACT_PROPOSALS` first, then use the output proposals for negative chips mining.

**Comment by mahyarnajibi** [comment]
Just for the reference, an all-in-one script is provided here: ```scripts/train_neg_props_and_sniper.sh``` which performs all steps including generating the negative chips and training the SNIPER detector afterward.
